### PR TITLE
aries-vcx backchannel

### PR DESCRIPTION
**Pull request by mirgee**
Signed-off-by: Miroslav Kovar <miroslavkovar@protonmail.com>

### CODE REVIEW

File: aries-backchannels/aries-vcx/resource/indypool.txn [hunk 1]
```diff
@@ -0,0 +1,4 @@
...
+{"reqSignature":{},"txn":{"data":{"data":{"alias":"Node4","blskey":"2zN3bHM1m4rLz54MJHYSwvqzPchYp8jkHswveCLAEJVcX6Mm1wHQD1SkPYMzUDTZvWvhuE6VNAkK3KxVeEmsanSmvjVkReDeBEMxeDaayjcZjFGPydyey1qxBHmTvAnBKoPydvuTAqx5f7YNNRAdeLmUi99gERUU7TD8KfAa6MpQ9bw","blskey_pop":"RPLagxaR5xdimFzwmzYnz4ZhWtYQEj8iR5ZU53T2gitPCyCHQneUn2Huc4oeLd2B2HzkGnjAff4hWTJT6C7qHYB1Mv2wU5iHHGFWkhnTX9WsEAbunJCV2qcaXScKj4tTfvdDKfLiVuU2av6hbsMztirRze7LvYBkRHV3tGwyCptsrP","client_ip":"127.0.0.1","client_port":9708,"node_ip":"127.0.0.1","node_port":9707,"services":["VALIDATOR"]},"dest":"4PS3EDQ3dW1tci1Bp6543CfuuebjFrg36kLAUcskGfaA"},"metadata":{"from":"TWwCRQRZ2ZHMJFn9TzLp7W"},"type":"0"},"txnMetadata":{"seqNo":4,"txnId":"aa5e817d7cc626170eca175822029339a444eb0ee8f0bd20d3b0b76e566fb008"},"ver":"1"}
```
Comment (ianco) [inline aries-backchannels/aries-vcx/resource/indypool.txn:4]: Note that there is a ledger url parameter passed to the `manage` script:

`export LEDGER_URL_CONFIG="${LEDGER_URL_CONFIG}"`

The pool transactions can be downloaded from `${LEDGER_URL_CONFIG}/genesis` - using this is a good way to keey the tests ledger agnostic

File: aries-backchannels/aries-vcx/src/setup.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
+use vcx::init::{open_pool, init_issuer_config, init_agency_client, open_pool_directly};
+use vcx::libindy::utils::wallet::{create_wallet_from_config, configure_issuer_wallet, open_wallet_directly, close_main_wallet};
+use vcx::utils::provision::provision_cloud_agent;
+use vcx::libindy::utils::pool;
+use std::io::prelude::*;
+use uuid;
+
+async fn download_tails() -> std::io::Result<std::path::PathBuf> {
+    info!("Downloading tails file");
+    let body = reqwest::get("http://127.0.0.1:9000/genesis")
```
Comment (ianco) [inline aries-backchannels/aries-vcx/src/setup.rs]: See the previous note about ${LEDGER_URL_CONFIG}

File: aries-backchannels/aries-vcx/src/setup.rs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        "wallet_name": format!("rust_agent_{}", uuid::Uuid::new_v4().to_string()),
+        "wallet_key": "8dvfYSt5d1taSd6yJdpjq4emkwsPDDLYxkNFysFD2cZY",
+        "wallet_key_derivation": "RAW",
+    }).to_string();
+    
+    create_wallet_from_config(&wallet_config).unwrap();
+    let _wh = open_wallet_directly(&wallet_config).unwrap();
+    let _ph = open_pool_directly(&json!({ "genesis_path": genesis_path.to_str() }).to_string()).unwrap();
+
+    let enterprise_seed = "000000000000000000000000Trustee1";
```
Comment (ianco) [inline aries-backchannels/aries-vcx/src/setup.rs]: DIDs can be registered through the von-network web server, POST to `${LEDGER_URL_CONFIG}/register`

Something like:

`curl -d '{\"seed\":\"${INDY_WALLET_SEED}\", \"role\":\"TRUST_ANCHOR\", \"alias\":\"${AGENT_NAME}\"}' -X POST ${LEDGER_URL}/register`

(alternately you can specify DID and verkey instead of seed)

(avoids having to hard-code the trustee seed, because the von network web browser knows it)

**Comment by swcurran** [comment]
Hi @mirgee -- is this PR ready to merge.  I thought it was a work in process, but don't see that.  Would love to add it to the regular runs and published results.

**Comment by mirgee** [comment]
Hi @swcurran - we have several passing scenarios in all of connection, issuance and proof presentation, but there still are things we would like to improve on the aries-vcx side (out of band, public dids, credential and proof proposals, ...). I was thinking we'll leave this open until we are done with those, but those are quite big features which may take some time, so I will put a few finishing touches and let's merge this (I will remove WIP once ready).

**Comment by nodlesh** [comment]
Hi @mirgee Just wanted to check in and see how things were going on the finishing touches on the VCX Backchannel? Hope it is going well. We can't wait to add this to the daily test executions. 👍

**Comment by mirgee** [comment]
@nodlesh Apologies, was busy with other things, but this PR is almost ready, except for the fact that there couple of changes to RFC0037 proof presentation steps related to #259 - either the other backchannels have to be updated to reflect the changes in the proof request / presentation format, or this one has to be updated and the changes reverted.

**Comment by nodlesh** [comment]
No changes have been made to AATH based on #259 yet. Let's cross the backchannel updates bridge when we get to 259.  

Otherwise is this ready for review?

**Comment by mirgee** [comment]
@nodlesh Yes, as I said, this is ready for review except for the changes made to the proof request / presentation format (#259) which need to be reconciled.

**Comment by nodlesh** [comment]
Should I be able to run this locally? Getting errors on build. 

@swcurran What should we do with the changes in this PR for #259? If we merge this as is it will cause the other agents to fail the daily run.   Maybe we should remove these updates and put them in a separate PR and do all the backchannel edits there?

File: .github/workflows/test-harness-ariesvcx-ariesvcx.yml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      - name: run-von-network
+        uses: ./test-harness/actions/run-von-network
+      - name: run-indy-tails-server
+        uses: ./test-harness/actions/run-indy-tails-server
+      - name: run-test-harness-wo-reports
+        uses: ./test-harness/actions/run-test-harness-wo-reports
+        with:
+          BUILD_AGENTS: "-a aries-vcx"
+          TEST_AGENTS: "-d aries-vcx"
+          TEST_SCOPE: "-t ~@T001-RFC0036 -t ~@T001.1-RFC0036 -t ~@T002-RFC0036 -t ~@T004-RFC0036 -t ~@T005-RFC0036 -t ~@T006-RFC0036 -t ~@T001.1-RFC0037 -t ~@T001.2-RFC0037 -t ~@T001.3-RFC0037 -t ~@T001.4-RFC0037 -t ~@T001.5-RFC0037 -t ~@T002-RFC0037 -t ~@T003-RFC0037 -t ~@T003.1-RFC0037 -t ~@T004-RFC0037 -t ~@T005-RFC0037 -t ~@T006-RFC0037 -t ~@T004-RFC0160 -t ~@T005-RFC0160 -t ~@T006-RFC0160 -t ~@T007-RFC0160 -t ~@RFC0023 -t ~@revocation -t ~@RFC0453 -t ~@RFC0454"
```
Comment (nodlesh) [inline .github/workflows/test-harness-ariesvcx-ariesvcx.yml]: It may be clearer to state the test scope as this below, which is equivalent.
`-t @T003-RFC0036,@T001-RFC0037,@T001-RFC0160,@T002-RFC0160,@T003-RFC0160`

**Comment by mirgee** [comment]
@nodlesh Yes, it should be possible to run it locally - I have no errors on build, and tested it on github actions as well.

As for the 259 updates, I may remove them and make the aries-vcx backchannel not rely on them.

**Comment by nodlesh** [comment]
This is what I'm getting on build.
```
...
++ echo
+ docker build -t aries-vcx-agent-backchannel -f aries-backchannels/aries-vcx/Dockerfile.aries-vcx aries-backchannels/
[+] Building 0.5s (3/3) FINISHED                                                                                                                                     
 => [internal] load build definition from Dockerfile.aries-vcx                                                                                                  0.0s
 => => transferring dockerfile: 47B                                                                                                                             0.0s
 => [internal] load .dockerignore                                                                                                                               0.0s
 => => transferring context: 35B                                                                                                                                0.0s
 => ERROR [internal] load metadata for docker.pkg.github.com/hyperledger/aries-vcx/libvcx:0.17.3                                                                0.4s
------
 > [internal] load metadata for docker.pkg.github.com/hyperledger/aries-vcx/libvcx:0.17.3:
------
failed to solve with frontend dockerfile.v0: failed to create LLB definition: unexpected status code [manifests 0.17.3]: 401 Unauthorized
+ echo 'Docker image build failed.'
Docker image build failed.
+ exit 1
```

**Comment by mirgee** [comment]
@nodlesh You have to [login](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#authenticating-to-the-container-registry) to github docker registry.

**Comment by nodlesh** [comment]
Even when I'm logged in to the GitHub container registry I still get the same issue. What rights does my token need? I gave it the `repo` scope.

**Comment by nodlesh** [comment]
I am also trying to run with the other agents to see if there are any negative side effects from these updates. Can't seem to build acapy from your branch for some reason. You might want to merge with the latest in main.  Also, I'm assuming the Cloud Agency is only for the vcx agent, it shouldn't be started or checked if running when other agents don't need it. Running with acapy only under normal conditions the manage script now gives an error that the cloud agency is not running.

**Review by ianco — COMMENTED** [review]
> ... Also, I'm assuming the Cloud Agency is only for the vcx agent, it shouldn't be started or checked if running when other agents don't need it. Running with acapy only under normal conditions the manage script now gives an error that the cloud agency is not running.

Is it possible to run the agency as part of the backchannel?  I.e. when the vcx agent is started, the cloud agency runs as a sub-process within the agent docker container?  As Sheldon mentions we're trying to keep the top-level manage script generic and embed the agent-specific dependencies within the backchannel.

File: aries-backchannels/aries-vcx/resource/indypool.txn (hunk 1 above)
Comment (mirgee) [reply aries-backchannels/aries-vcx/resource/indypool.txn:4]: I know and the backchannel does download the von node genesis file. This genesis file is useful for development while running against indypool locally (not inside a docker container).

File: aries-backchannels/aries-vcx/src/setup.rs (hunk 2 above)
Comment (mirgee) [reply aries-backchannels/aries-vcx/src/setup.rs]: You are reviewing an old version of the PR.

Comment (ianco) [reply aries-backchannels/aries-vcx/src/setup.rs]: Yah I made this comment weeks ago not sure why it's just showing up now ...

**Comment by mirgee** [comment]
> Is it possible to run the agency as part of the backchannel? I.e. when the vcx agent is started, the cloud agency runs as a sub-process within the agent docker container?

@ianco This is not impossible, but unfortunately would be quite cumbersome - the agency must have unique address from the POV of any participant (i.e. when multiple aries-vcx backchannel agents are running, each with its own agency, their respective agencies must have the same address for the agent in whose container they run in and all other agents).

But obviously, the cloud agency does not need to run if none of the agents is of aries-vcx type (although running it does not cause any issues).

File: aries-backchannels/aries-vcx/src/setup.rs (hunk 3 above)
Comment (mirgee) [reply aries-backchannels/aries-vcx/src/setup.rs]: Good to know, thanks.

**Comment by mirgee** [comment]
@nodlesh The link I provided you with names exactly the permissions you need. If you are really desperate, try updating your docker client (see [this](https://github.com/docker/compose-cli/issues/1545)).

**Comment by mirgee** [comment]
Hi @nodlesh , have you been able to fix the issue and test this backchannel?

**Comment by nodlesh** [comment]
@mirgee Issue #259 has been merged. Would you be able to adjust this PR to take the changes in proof request/proposal data?

File: aries-test-harness/features/steps/0160-connection.py [hunk 5]
```diff
@@ -118,7 +118,7 @@ def step_impl(context, invitee):
 
     resp_json = json.loads(resp_text)
 
-    if not hasattr(context, 'connection_id_dict'):
+    if not hasattr(context, 'connection_id_dict') or not hasattr(context.connection_id_dict, invitee):
```
Comment (Moopli) [inline aries-test-harness/features/steps/0160-connection.py]: This will replace the connection_id_dict with a fresh one if the invitee is not present, which loses what else might have been there.

Adding a missing invitee dict should be outside, eg:
```
if not hasattr(context, 'connection_id_dict'):
    context.connection_id_dict = {}
context.connection_id_dict.setdefault(invitee, {})
```

Comment (mirgee) [reply aries-test-harness/features/steps/0160-connection.py]: Agreed, done

**Comment by mirgee** [comment]
@nodlesh I adjusted the format and ran the tests.

File: .github/workflows/test-harness-ariesvcx-ariesvcx.yml (hunk 4 above)
Comment (nodlesh) [reply .github/workflows/test-harness-ariesvcx-ariesvcx.yml]: @mirgee Can you change the Test Scope to the tests you are running instead of what is being excluded?

File: manage [hunk 6]
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-# set -x
+set -x
```
Comment (nodlesh) [inline manage]: Please comment set -x out.

**Comment by nodlesh** [comment]
I'm having an issue building the agent.  @mirgee Can you help?
Command to build
`./manage rebuild -a aries-vcx`

Error: (full output from build attached below)
```
#12 406.5 error: linking with `cc` failed: exit status: 1
#12 406.5   |
#12 406.5   = note: "cc" "-m64" "/home/node/home/agent/aries-vcx-backchannel/aries-vcx/target/debug/deps/aries_vcx_backchannel-9d60210664686cfa.11o8hfc4oupu0xga.rcgu.o" <long series of output here>
#12 406.5   = note: collect2: fatal error: ld terminated with signal 9 [Killed]
#12 406.5           compilation terminated.
#12 406.5           
#12 406.5 
#12 407.0 warning: `aries-vcx-backchannel` (bin "aries-vcx-backchannel") generated 69 warnings
#12 407.8 error: could not compile `aries-vcx-backchannel` due to previous error; 69 warnings emitted
```
[build-aries-vcx.txt](https://github.com/hyperledger/aries-agent-test-harness/files/7295942/build-aries-vcx.txt)

**Comment by mirgee** [comment]
@nodlesh That's weird, I am not getting this error. Are you sure you are using the latest revision of the branch with no other changes?

**Comment by nodlesh** [comment]
Yes, I am using the latest, unmodified. I even recloned and ran again. Same issue.

**Comment by mirgee** [comment]
@nodlesh Can you try increasing the memory limit of the docker client (e.g. I am using 8 GB)?

**Comment by swcurran** [comment]
I ran it and it worked fine for me.  Linux machine with lots of memory.

File: aries-backchannels/.dockerignore [hunk 7]
```diff
@@ -1,4 +1,16 @@
 # This is needed because indy-sdk needs to be installed
 # specifically for the platform it is installed. So we can't
 # mount the node modules from the host
-javascript/server/node_modules
+javascript/**
```
Comment (nodlesh) [inline aries-backchannels/.dockerignore]: The contents of this file is causes all other agents to fail to build. Can you revert the changes here?

File: aries-test-harness/features/steps/0036-issue-credential.py [hunk 8]
```diff
@@ -271,6 +272,11 @@ def step_impl(context, holder):
     
     assert resp_status == 200, f'resp_status {resp_status} is not 200; {resp_text}'
     resp_json = json.loads(resp_text)
+    resp_state = resp_json["state"]
```
Comment (nodlesh) [inline aries-test-harness/features/steps/0036-issue-credential.py]: These lines (State checking) were removed in later versions of issue credential. You should merge with the latest and do not add these back in.

File: aries-test-harness/features/steps/0160-connection.py [hunk 9]
```diff
@@ -120,16 +120,15 @@ def step_impl(context, invitee):
...
+
     
     context.connection_id_dict[invitee][context.inviter_name] = resp_json["connection_id"]
 
     # Also add the inviter into the main connection_id_dict. if the len is 0 that means its already been cleared and this may be Mallory.
     if "temp_connection_id_dict" in context:
         if len(context.temp_connection_id_dict) != 0:
             context.connection_id_dict[context.inviter_name] = {invitee: context.temp_connection_id_dict[context.inviter_name]}
-            #clear the temp connection id dict used in the initial step. We don't need it anymore.
-            context.temp_connection_id_dict.clear()
```
Comment (nodlesh) [inline aries-test-harness/features/steps/0160-connection.py]: Why was this removed?

File: aries-test-harness/features/steps/0160-connection.py [hunk 10]
```diff
@@ -430,7 +429,7 @@ def step_impl(context, inviteinterceptor, inviter):
 @then(u'"{inviter}" sends a request_not_accepted error')
 def step_impl(context, inviter):
     inviter_url = context.config.userdata.get(inviter)
-    inviter_connection_id = context.connection_id_dict[inviter][context.invitee_name]
+    inviter_connection_id = context.connection_id_dict[inviter][context.inviteinterceptor_name]
```
Comment (nodlesh) [inline aries-test-harness/features/steps/0160-connection.py]: Why was this changed to the inviteeInterceptor?

**Comment by nodlesh** [comment]
Got the aries-vcx tests running and passing. 👍
A few things need updating. Aside from my comments on some of the files directly, I'm not sure you have the latest from main? There are recent edits to the manage script that you do not have.  You should merge with main and update the rest of the files accordingly. Thanks.

**Comment by mirgee** [comment]
@nodlesh You are right, some of the latest changes were missing. Aside from merging in the latest master (done) it seems no other changes are necessary.

**Comment by nodlesh** [comment]
My reviews are in the files themselves, along with the request for changes. There should be a clearer -t change, removal of the changes in .dockerignore file (causes other agent build failure), commenting out the `set -x` in the manage file, along with some explanations on the changes in connections.py and issue_credential.py.  Are you not seeing those comments I've made?

**Comment by mirgee** [comment]
@nodlesh No, I am not seeing your comments.

**Review by nodlesh — CHANGES_REQUESTED** [review]

**Comment by nodlesh** [comment]
Sorry, my comments on the files should be there now.

File: aries-test-harness/features/steps/0160-connection.py (hunk 9 above)
Comment (mirgee) [reply aries-test-harness/features/steps/0160-connection.py]: I don't remember why I made these changes - the concerned scenarios pass without them, so I reverted them.

File: aries-test-harness/features/steps/0160-connection.py (hunk 10 above)
Comment (mirgee) [reply aries-test-harness/features/steps/0160-connection.py]: I don't remember why I made these changes - the concerned scenarios pass without them, so I reverted them.

**Comment by nodlesh** [comment]
This is looking good. No side effects on other agent runsets. Thanks for your hard work @mirgee

**Review by nodlesh — APPROVED** [review]
